java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileWriter.java

```diff
@@ -130,14 +130,24 @@ protected void endInternal(WriteChannel out) throws IOException {
   protected void ensureDictionariesWritten(DictionaryProvider provider, Set<Long> dictionaryIdsUsed)
       throws IOException {
     if (dictionariesWritten) {
+      for (long id : dictionaryIdsUsed) {
+        BaseDictionary dictionary = provider.lookup(id);
+        if (dictionary.getEncoding().isDelta()) {
+          writeDictionaryBatch(dictionary, false);
+        } else if (dictionary.mark() > 0) {
+          throw new IllegalStateException("Replacement dictionaries are not supported in the " +
+              "IPC file format. Dictionary ID: " + dictionary.getEncoding().getId());
+        }
+      }
       return;
     }
+
     dictionariesWritten = true;
     // Write out all dictionaries required.
     // Replacement dictionaries are not supported in the IPC file format.
     for (long id : dictionaryIdsUsed) {
-      Dictionary dictionary = provider.lookup(id);
-      writeDictionaryBatch(dictionary);
+      BaseDictionary dictionary = provider.lookup(id);
+      writeDictionaryBatch(dictionary, true);
```

Comment:
nit: may be comment the variable name for `true`

Reply:
👍